webidl.txt

```diff
@@ -148,51 +173,74 @@ dictionary VideoEncodeLayer {
   unsigned long? bitsPerSecond;
 }
 
-[Constructor(BufferSource data, MediaTime timestamp)]
-interface EncodedVideoFrame {
-  readonly attribute Uint8Array data;
-  readonly attribute MediaTime timestamp;
-  // Info provided as a result from the encoder
-  // Not needed as input to a decoder
-  readonly attribute VideoEncodeResult? encoded;
+dictionary VideoEncoderStaticSettings {
+  // Can be used to initialize the encoder faster
+  // than waiting for the first frame
+  unsigned long? expectedWidth;
+  unsigned long? expectedHeight;  
+
+  sequence<VideoEncodeLayer> layers;
+}
+
+dictionary VideoEncoderDynamicSettings {
+  // unset/null means the encoder will pick
+  // target will be exceeded for key frames
+  unsigned long long? targetBitRate;
+
+  VideoEncodeContentMode contentMode;
+}
+
+dictionary VideoEncoderSettings : VideoEncoderStaticCodecSettings, VideoEncoderDynamicCodecSettings {
+}
+
+dictionary VideoEncoderInit : VideoCodecParameters {
+  VideoEncoderSettings settings;
 }
 
-interface VideoEncodeResult {
+dictionary VideoEncoderFrameSettings {
+  boolean? forceKeyFrame;
+}
+
+dictionary VideoEncoderInput {
+  VideoFrame frame;
+  VideoEncoderFrameSettings frameSettings;
+  VideoEncoderDynamicSettings changeCodecSettings;
+}
+
+dictionary VideoEncoderOutput {
+  Uint8Array data;
```

Comment:
I'm not sure it's good to switch from interfaces to dictionaries with (passive) data at this point. If we want to make efficient transformations that only look at the metadata (such as a layer-discarding filter), it may be better to leave the data as an accessor so that the case of passing frames unmodified can be handled without any data copying.

Reply:
This is already the case for the heavyweight resources (e.g., raw video frames) but not for bitstream contents. I think that discussion is outside the scope of this PR but would be valuable.